common/tasks/interleaved_weighted_round_robin.go

```diff
@@ -154,8 +150,7 @@ func (s *InterleavedWeightedRoundRobinScheduler[T, K]) TrySubmit(
 	task T,
 ) bool {
 	numTasks := atomic.AddInt64(&s.numInflightTask, 1)
-	if numTasks == 1 {
-		s.doDispatchTasksDirectly(task)
+	if numTasks == 1 && s.tryDispatchTaskDirectly(task) {
```

Comment:
i thought this PR is to remove this optimization?

Reply:
Removing this optimization is my quick hack for testing multi cursor implementation.
This optimization does make sense to me so I don't want to simply remove it. 

The effort to keep it and fix the issue is not huge and most of this PR is actually doing refactoring for consolidating `Scheduler` and `Processor` interface after adding `TrySubmit` to `Processor` interface.